text/0000-std-aware-cargo.md

```diff
@@ -0,0 +1,289 @@
+- Feature Name: cargo_the_std_awakens
+- Start Date: 2018-02-09
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Currently, the `core` and `std` components of Rust are handled in a different way than Cargo handles other crate dependencies. This causes issues for non-mainstream targets, such as WASM, Embedded, and new not-yet-tier-1 targets. The following RFC proposes a roadmap to address these concerns in a consistent and incremental process.
+
+# Motivation
+[motivation]: #motivation
+
+In today's Rust environment, `core` and `std` are shipped as precompiled objects. This was done for a number of reasons, including faster compile times, and a more consistent experience for users of these dependencies. This design has served fairly well for the bulk of users, however there are a number of less common uses of Rust, that are not well served by this approach. Examples include:
+
+* Supporting new/arbitrary targets, such as those defined by a ".json" file
+* Making modifications to `core` or `std` through use of feature flags
+* Users who would like to make different optimizations to `core` or `std`, such as `opt-level = 'z'`, with `panic = "abort"`
```

Comment:
I'd add that we'd really like to be able to have debug_assertions in `core`, `alloc`, and `std`.  See threads like https://internals.rust-lang.org/t/make-vec-set-len-enforce-the-len-cap-invariant/8927?u=scottmcm

Reply:
Hey @scottmcm, is this something you would see as a compile time configuration option of `core`, `std`, `alloc`, etc?

This RFC isn't meant to propose which features should or should not be available, but instead is focused on the method for selecting them.

Reply:
@jamesmunns Yes, I do see it as a compile-time configuration option.  (Ditto for having overflow checks on by default in debug but not in release -- right now that's hacked together with `#[rustc_inherit_overflow_checks]`, which I'd love to be able to remove.)

Reply:
And I'm fine with the RFC not saying how those particular things should be accomplished, but I think they're valuable examples since it's the Motivation section.

Reply:
@scottmcm Would you be interested in PR-ing (or suggesting text) for a section in "Future possibilities" to discuss a listing of possible flags to stabilize? I don't think this should impact the RFC, but might be good to centralize a list of things people are thinking of.

In general, I want to make it clear in that section that "this is a list of possible flags to stabilize, but this RFC does not guarantee any of the following will be stabilized".

Feel free to fork my repo and submit a PR to this branch, and I can accept it there.